### PR TITLE
[components] Split out dg-cli from dagster-components

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Splits out `dg-cli` from `dagster-components`. `dg-cli` does not depend on any other dagster package.

All possible generation logic is moved to `dg-cli`. However, `dg-cli` sometimes needs to run commands in a code location environment:

- Querying the environment to see the defined components
- Running the `generate_files` routine to generate a new component

This functionality is exposed from `dagster-components` via a new `dagster-components` CLI, that is basically a backend for `dg`. It's not intended to be called by users. The plumbing between the two packages could be improved but this is a working first implementation.

Other comments:

- `dagster-components` no longer has any concept of a "deployment"-- the `DeploymentProjectContext` has been moved to `dg`. However, both `dg-cli` and `dagster-components` have slightly differing versions of `CodeLocationProjectContext`. The `dg-cli` version uses a new `RemoteComponentRegistry` and `RemoteComponentType` class to represent the contents of the component registry in the code location.
- `dg generate component` will forward the `-- extra args` or `--json-params` untouched onto `dagster-components generate component`
- `dagster-components` doesn't have any _direct_ CLI testing right now-- the testing is done via the higher level calls from `dg`

## How I Tested These Changes

Unit tests.

### CODE REVIEW

**Comment by smackesey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26364?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#26364** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26364?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26364?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#26295** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26295?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#26337](https://github.com/dagster-io/dagster/pull/26337) <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/26337?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* `master`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: python_modules/libraries/dg-cli/dg_cli/cli/generate.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    code_location_path = os.path.join(context.code_location_root_path, name)
+
+    generate_code_location(code_location_path, editable_dagster_root)
+
+
+@generate_cli.command(name="component-type")
+@click.argument("name", type=str)
+def generate_component_type_command(name: str) -> None:
+    """Generate a Dagster component instance."""
+    if not is_inside_code_location_project(Path(".")):
```
Comment (OwenKephart) [inline python_modules/libraries/dg-cli/dg_cli/cli/generate.py:79]: Is there a difference between Path(".") and Path.cwd()? If not, should we just stick with Path.cwd()?

**Review by OwenKephart — APPROVED** [review]
I actually quite like the setup where there's a code-location specific CLI that we hook into

File: python_modules/libraries/dg-cli/dg_cli/cli/generate.py (hunk 1 above)
Comment (smackesey) [reply python_modules/libraries/dg-cli/dg_cli/cli/generate.py:79]: I don't think there's a difference. Agree that `Path.cwd()` is preferred. Will change.

**Pull request merged**